Userland/Utilities/traceroute.cpp

```diff
@@ -0,0 +1,175 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/String.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/ElapsedTimer.h>
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <netinet/ip_icmp.h>
+#include <serenity.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio id inet unix", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    const char* host_name;
+    int max_hops = 30;
+    int max_retries = 3;
+
+    Core::ArgsParser args_parser;
+    args_parser.add_positional_argument(host_name, "destination", "destination", Core::ArgsParser::Required::Yes);
+    args_parser.add_option(max_hops, "use at most <hops> to the destination", "max-hops", 'h', "hops");
+    args_parser.add_option(max_retries, "retry TTL at most <tries> times", "max-tries", 't', "tries");
+    args_parser.parse(argc, argv);
+
+    if (max_hops < 1 || max_hops > 255) {
+        warnln("Invalid maximum hops amount");
+        return 1;
+    }
+
+    if (max_retries < 1) {
+        warnln("Invalid maximum retries amount");
+        return 1;
+    }
+
+    auto* hostent = gethostbyname(host_name);
+    if (!hostent) {
+        warnln("Lookup failed for '{}'", host_name);
+        return 1;
+    }
+    sockaddr_in host_address {};
+    memset(&host_address, 0, sizeof(host_address));
+    host_address.sin_family = AF_INET;
+    host_address.sin_port = 44444;
+    host_address.sin_addr.s_addr = *(const in_addr_t*)hostent->h_addr_list[0];
+
+    int fd = socket(AF_INET, SOCK_RAW, IPPROTO_ICMP);
+    if (fd < 0) {
+        perror("socket");
+        return 1;
+    }
+
+    if (setgid(getgid()) || setuid(getuid())) {
+        warnln("Failed to drop privileges");
+        return 1;
+    }
+
+    if (pledge("stdio inet unix", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    timeval timeout { 5, 0 };
+    if (setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout)) < 0) {
+        perror("setsockopt");
+        return 1;
+    }
+
+    struct icmp_request {
+        struct icmphdr header;
+        char msg[64 - sizeof(struct icmphdr)];
+    };
+
+    struct icmp_response {
+        u8 ip_header[20];
+        struct icmphdr header;
+        char msg[64 - sizeof(struct icmphdr)];
+    };
+
+    // 1: reached target
+    // 0: got ttl exhausted response
+    // -1: error or no response
+    auto try_reach_host = [&](int ttl) -> int {
+        Core::ElapsedTimer m_timer { true };
+        for (auto i = 0; i < max_retries; i++) {
+            icmp_request request {};
+            request.header = { ICMP_ECHO, 0, 0, { { 0, 0 } } };
+            VERIFY(String::number(ttl).copy_characters_to_buffer(request.msg, sizeof(request.msg)));
```

Comment:
I know it's not kosher to do things with side effects in an assertion, but does the same rule apply to VERIFY?

Reply:
It technically doesn't apply since `VERIFY` never gets compiled out.
However, it looks really odd so let's not do it :)